### DISCUSSION TITLE
Change of the parameter slope in version 1.5 breaks older models

### DISCUSSION THREAD

**Discussion by CaioDaumann** [probabilists/zuko#74] [General]
Dear Zuko experts,

It may have come to your attention that your library is currently being used by the CMS experiment to perform corrections for photon precision measurements, including the Higgs boson decaying into the diphoton channel (see the current results: https://arxiv.org/abs/2504.17755).

Recently, we ran into a problem with the new Zuko version, in which the slope parameter is now trained. As a result, the models we previously trained (zuko.flows.NSF) no longer work with the new version, presumably due to the change in the slope parameter (https://github.com/probabilists/zuko/compare/1.4.1...1.5.0#diff-ebd6c72181891f49eb19248d9a7a211dd76012e45796d35cd8348c12941aaef8L442).
 Therefore, I would like to ask whether it would be possible to roll back this change so that models trained previously remain compatible with version 1.5 or future releases. This would be a good solution for us, as new data keeps coming in and we need to continuously train new models.

I could also copy the relevant function and manually set the parameter. But I would prefer to use the functions/classes directly from the Zuko package.  Please let me know what you think would be the best approach.

And what was motivation for the parameter change? Does it improve the model performance? What would be your recommendation when training new models? 

Cheers,
Caio

**Comment by francois-rozet** [comment]
Hi Caio, thank you for opening this discussion. I was not aware that Zuko was used by the CMS collaboration, but I am very happy and honored to learn it!

I am sorry for the issues caused by the latest release. I want to fix this as soon as possible, but first let me explain why I made this change (in https://github.com/probabilists/zuko/commit/f498c507c121b6226729a392c3e0cbbd4a7ddb8f).

In theory, a normalizing flow is a composition of invertible and differentiable parameterized transformations. In practice, computers use floating-point representation to store real numbers, which makes some numerical algorithms (which are analytically exact) unstable. In our case, when the slope of a transformation is very close to 0, it is hard to guarantee precise numerical inversion. You find similar issues with inverting a matrix with very small eigenvalues.

With the minimum slope we used previously (1e-4), we had reports of numerical instability, especially in single-precision (`float32`). We wanted to fix this issue by enforcing a tighter threshold (1e-3), which worked well. I believe that this was a good choice overall, but indeed, we should have informed the community of this change.

The solution I propose would be to provide an easy way for the user to roll back to the legacy slope parameter, either with a flag to activate or by giving more control of these parameters in the flow constructor(s). Let me know what you think!

**Reply by CaioDaumann** [reply]
Hi @francois-rozet ,

Thanks for the proposed solution. Indeed, a flag or another way to use the old value would be extremely useful. Then, if a new model is trained with a different slope parameter, we can easily set it accordingly. Thanks!

**Reply by francois-rozet** [reply]
Hi @CaioDaumann, I have implemented a solution in #77. Could you please check if it suits your needs?

**Reply by CaioDaumann** [reply]
Hi @francois-rozet , it looks exactly what we need, thanks! 

In order to change the slope, I would only need to do this, right? 

```
flow_model = zuko.flows.NSF(
    13,
    5,
    bins=10,
    transforms=5,
    hidden_features=[256] * 2,
    slope=1e-4
    )
```

**Reply by francois-rozet** [reply]
Yes, it should be as simple as what you wrote. Maybe you can try it out?

```
pip install git+https://github.com/probabilists/zuko@legacy-slope
```

**Reply by CaioDaumann** [reply]
Alright! @francois-rozet  I was having a try and I am getting a different error now, related to the name change of some keys:

from: `"base._0", "base._1"` to `"base.loc", "base.scale"`.

had a (quick) look at the differences between versions 1.4.1 and 1.5.0, and also at your new PR, but I don’t see any changes to these parameter names. Any idea why this might be happening?

**Reply by CaioDaumann** [reply]
I just checked with zuko 1.5.0 `pip install zuko==1.5.0` and I dont get this error. Only when using:

`pip install git+https://github.com/probabilists/zuko@legacy-slope`

**Reply by francois-rozet** [reply]
Oh yes, that's the arguments of the base distribution, which were changed to be keyword arguments rather than positional arguments. This is safer and easier to debug. You should be able to fix this with `strict=False` in `load_state_dict`.

I would recommend doing a quick load -> load_state_dict (strict=False) -> state_dict -> save to get "translated" version of the checkpoint.

> Note this was not introduced by the PR, but by an older commit, just after the release of 1.5.0

**Reply by CaioDaumann** [reply]
Ok, thanks for the suggestion @francois-rozet . I ran and validated everything on our side with the PR. With the changes introduced in https://github.com/probabilists/zuko/pull/77, the flows now work as expected. Thanks!

**Reply by francois-rozet** [reply]
Great, thanks! The PR is now merged, it should be part of the next release soon.
